src/rvpt/bvh_builder.cpp

```diff
@@ -0,0 +1,183 @@
+﻿//
+// Created by legend on 21/8/2020.
+//
+
+#include "bvh_builder.h"
+
+#include <numeric>
+#include <algorithm>
+#include <cassert>
+
+Bvh BinnedBvhBuilder::build_bvh(const std::vector<glm::vec3>& primitive_centers,
+                                const std::vector<AABB>& bounding_boxes)
+{
+    assert(primitive_centers.size() == bounding_boxes.size());
+    size_t primitive_count = primitive_centers.size();
+    assert(primitive_count != 0);
+
+    // Allocate enough nodes in a big, flat array. For this, we use the property:
+    //
+    //   number of nodes = 2 * number of leaves - 1
+    //
+    // This property is valid for every binary tree and easy to verify by induction.
+    // Since a BVH has at most as many leaves as there are primitives (one primitive
+    // only goes into one leaf), then we have an upper bound on the number of nodes.
+    Bvh bvh;
+    bvh.nodes.reserve(2 * primitive_count - 1);
+
+    // Primitive indices are just a big array of indices into the primitive data.
+    // Initially, that array is just 0, 1, 2, ...
+    bvh.primitive_indices.resize(primitive_count);
+    std::iota(bvh.primitive_indices.begin(), bvh.primitive_indices.end(), 0);
+
+    // Initially, we set the root node to be a leaf that spans the entire list of primitives
+    bvh.nodes.emplace_back(BvhNode{.first_child_or_primitive = 0,
+                                   .primitive_count = static_cast<uint32_t>(primitive_count)});
+    build_bvh_node(bvh, bvh.nodes.front(), primitive_centers, bounding_boxes);
+    bvh.nodes.shrink_to_fit();
+    return bvh;
+}
+
+size_t BinnedBvhBuilder::compute_bin_index(int axis, const glm::vec3& center,
+                                           const AABB& aabb) noexcept
+{
+    int index =
+        (center[axis] - aabb.min[axis]) * (static_cast<float>(bin_count) / aabb.diagonal()[axis]);
+    return std::min(int{bin_count - 1}, std::max(0, index));
+}
+
+std::tuple<float, int, size_t> BinnedBvhBuilder::find_best_split(
```

Comment:
Return a struct instead of a tuple, since there isn't anything to be gained by using a tuple, and increases compile time while reducing readability (users have to know what the float,int, and size_t represent).